src/main/cpp/src/case_when.cu

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "case_when.hpp"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/strings/detail/strings_column_factories.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+
+#include <thrust/transform.h>
+
+namespace spark_rapids_jni {
+namespace detail {
+
+/**
+ * Select the column index for the first true in bool columns for the specified row
+ */
+struct select_first_true_fn {
+  // bool columns stores the results of executing `when` expressions
+  cudf::table_device_view const d_table;
+
+  /**
+   * The number of bool columns is the size of case when branches.
+   * Note: reuturned index may be out of bound, valid bound is [0, col_num)
+   * When returning col_num index, it means final result is NULL value or ELSE value.
+   *
+   * e.g.:
+   * CASE WHEN 'a' THEN 'A' END
+   *   The number of bool columns is 1
+   *   The number of scalars is 1
+   *   Max index is 1 which means using NULL(all when exprs are false).
+   * CASE WHEN 'a' THEN 'A' ELSE '_' END
+   *   The number of bool columns is 1
+   *   The number of scalars is 2
+   *   Max index is also 1 which means using else value '_'
+   */
+  __device__ cudf::size_type operator()(std::size_t row_idx)
+  {
+    auto col_num                     = d_table.num_columns();
+    bool found_true                  = false;
+    cudf::size_type first_true_index = col_num;
+    for (auto col_idx = 0; !found_true && col_idx < col_num; col_idx++) {
+      auto const& col = d_table.column(col_idx);
+      if (!col.is_null(row_idx) && col.element<bool>(row_idx)) {
+        // Predicate is true and not null
+        found_true       = true;
+        first_true_index = col_idx;
+      }
+    }
+    return first_true_index;
```

Comment:
```
    auto const it = thrust::find_if(thrust::seq, d_table.begin(), d_table.end(), [row_idx]__device__(auto const& col) {
        return !col.is_null(row_idx) && col.element<bool>(row_idx);
    });
    return thrust::distance(d_table.begin(), it); // if not found, this will return `d_table.num_columns()`
```
Got this building error:
[exec] /home/chongg/code/spark-rapids-jni/src/main/cpp/src/case_when.cu(55): error: type name is not allowed.

And I saw thrust::find_if is complicated:
```
template<typename DerivedPolicy, typename InputIterator, typename Predicate>
_CCCL_HOST_DEVICE
InputIterator find_if(thrust::execution_policy<DerivedPolicy> &exec,
                      InputIterator first,
                      InputIterator last,
                      Predicate pred)
{
  typedef typename thrust::iterator_traits<InputIterator>::difference_type difference_type;
  typedef typename thrust::tuple<bool,difference_type> result_type;

  // empty sequence
  if(first == last) return last;

  const difference_type n = thrust::distance(first, last);

  // this implementation breaks up the sequence into separate intervals
  // in an attempt to early-out as soon as a value is found

  // TODO incorporate sizeof(InputType) into interval_threshold and round to multiple of 32
  const difference_type interval_threshold = 1 << 20;
  const difference_type interval_size = (thrust::min)(interval_threshold, n);

  // force transform_iterator output to bool
  typedef thrust::transform_iterator<Predicate, InputIterator, bool> XfrmIterator;
  typedef thrust::tuple<XfrmIterator, thrust::counting_iterator<difference_type> > IteratorTuple;
  typedef thrust::zip_iterator<IteratorTuple> ZipIterator;

  IteratorTuple iter_tuple = thrust::make_tuple(XfrmIterator(first, pred),
                                                thrust::counting_iterator<difference_type>(0));

  ZipIterator begin = thrust::make_zip_iterator(iter_tuple);
  ZipIterator end   = begin + n;

  for(ZipIterator interval_begin = begin; interval_begin < end; interval_begin += interval_size)
  {
    ZipIterator interval_end = interval_begin + interval_size;
    if(end < interval_end)
    {
      interval_end = end;
    } // end if

    result_type result = thrust::reduce(exec,
                                        interval_begin, interval_end,
                                        result_type(false,interval_end - begin),
                                        find_if_functor<result_type>());

    // see if we found something
    if(thrust::get<0>(result))
    {
      return first + thrust::get<1>(result);
    }
  }

  //nothing was found if we reach here...
  return first + n;
}

Let's use current concise impl?

```